### ISSUE TITLE
[Improvement] Schedule

### ISSUE THREAD

**Issue by mioux** [mihai-dinculescu/tapo#166]
By using the official app, I have seen there is a scheduler to turn P100 off or on

It seems not to be a timer integrated in the app, as I tried to completely shut down my phone after setting the timer, the plug has switched on while the phone was powered off

If you need some tests to get the API used, I can try to help

**Comment by mihai-dinculescu** [comment]
You are correct. The schedule is stored on the device itself, and the Tapo App is not required to trigger it.

I have previously encountered the schedule set and get API endpoints, so I am confident that support for them can be added.

**Comment by mioux** [comment]
I tried to setup a wireshark to capture packet sent to the Tapo without success (i'm a beginner in network :D)

If you want me to capture this, it may take some time (I have successuly setup wireshark to decode https packets, I did not see one going to the tapo)

**Comment by mihai-dinculescu** [comment]
I don't think that Wireshark is needed.
All the information should be in the decompiled Android app.

I see the following methods available in the API:
- get_schedule_rules
- get_schedule_next_action
- get_schedule_day_runtime
- get_schedule_month_runtime
- add_schedule_rule
- edit_schedule_rule
- remove_schedule_rules
- remove_all_schedule_runtime

The easiest way to go about it would be to set different schedules in the Tapo App and use `get_schedule_rules` to see what the result looks like. That can then be followed by trying to get `add_schedule_rule` to create schedules that appear in the App as they would if they were natively created.

**Comment by tzven0** [comment]
Hey, i second this. Any chance this will be implemented soon?
Sadly, i am just starting out with rust and i am currently using the tapo-rest API which uses your tapo lib for the sake of convenience.
Can i help out somehow, although i am just starting out with r?
Any chance i can test stuff with postman, did you test this?

**Comment by Tkd-Alex** [comment]
I would like to automatically power off my plug  after 10 minutes of my computer shutdowns.
The schedule function could help me, or do you have any other ideas to send a delayed command?

**Comment by goofy2k** [comment]
Hi @mihai-dinculescu ,

Do I understand from this
```
All the information should be in the decompiled Android app.

I see the following methods available in the API
```

that the API for the plugs is available (probably via reverse engineering)?
Where can I find it?

**Comment by mihai-dinculescu** [comment]
I've managed to decompile it by using ApkStudio on the [downloaded Android APK](https://www.apkmirror.com/?post_type=app_release&searchtype=apk&bundles%5B%5D=apkm_bundles&bundles%5B%5D=apk_files&s=tapo).

**Comment by Tkd-Alex** [comment]
Hello @mihai-dinculescu if you are able to create a little walkthrough to implement a new feature on the library I could try to develop the schedule part.

**Comment by mihai-dinculescu** [comment]
I believe expanding the CONTRIBUTING.md guide is a fantastic idea, and I warmly welcome any contributions to it.

However, please keep in mind that the main challenge lies in understanding the API by examining the decompiled Java Android App. This is considerably more complex than grasping the workings of this library.

**Comment by NyrZ** [comment]
Hey, similarly to @Tkd-Alex I'm looking for a way to schedule my P110 plug to turn off 5mins after shutting down my computer.

Any news on the implementation of the schedules/timers ?

**Comment by Hueburtsonly** [comment]
@Tkd-Alex , @NyrZ  For your desired feature of scheduling a P100 plug to turn off X minutes after shutting down your computer, you actually want the Timer API (e.g. do something X minutes/hours from now), not the Schedule API (e.g. do something every 8am on weekdays, or 15 minutes after sunset every Wednesday). The Timer API was added by https://github.com/mihai-dinculescu/tapo/pull/575 -- so you folks are ready to go!
